File: core/indicators.py
"""
Módulo de indicadores técnicos.
Calcula ATR, medias móviles, Z-score, momentum, volatilidad y más.
Todos los cálculos usan numpy/pandas para eficiencia.
"""
from __future__ import annotations
from typing import Iterable, Optional
import numpy as np
import pandas as pd


class Indicators:
    """Calculadora de indicadores técnicos sobre series de precios."""

    @staticmethod
    def sma(series: pd.Series, period: int) -> pd.Series:
        """Simple Moving Average."""
        return series.rolling(window=period, min_periods=period).mean()

    @staticmethod
    def ema(series: pd.Series, period: int) -> pd.Series:
        """Exponential Moving Average."""
        return series.ewm(span=period, adjust=False).mean()

    @staticmethod
    def std(series: pd.Series, period: int) -> pd.Series:
        """Desviación estándar rolling."""
        return series.rolling(window=period, min_periods=2).std()

    @staticmethod
    def atr(
        high: pd.Series, low: pd.Series, close: pd.Series, period: int = 14
    ) -> pd.Series:
        """Average True Range — mide volatilidad."""
        prev_close = close.shift(1)
        # np.fmax, not np.maximum and not pd.concat().max(): on the first row tr2/tr3 are NaN and
        # pandas' max SKIPS them, so fmax is the one that reproduces it exactly. Building a
        # three-column frame per call cost 35 s of a 245 s backtest across 10,161 calls (2026-09-04).
        tr = np.fmax(np.fmax((high - low).to_numpy(dtype=float),
                             (high - prev_close).abs().to_numpy(dtype=float)),
                     (low - prev_close).abs().to_numpy(dtype=float))
        true_range = pd.Series(tr, index=close.index)
        # Wilder's smoothing: equivalent to EWM with span = 2*period - 1
        return true_range.ewm(span=2 * period - 1, adjust=False).mean()

    @staticmethod
    def zscore(series: pd.Series, period: int) -> pd.Series:
        """Z-score: cuántas desviaciones estándar del precio respecto a su media."""
        mean = series.rolling(window=period, min_periods=2).mean()
        std = series.rolling(window=period, min_periods=2).std()
        deviation = series - mean
        # Guard against near-zero std: values below epsilon produce extreme z-scores
        safe_std = std.where(std > 1e-12, np.nan)
        result = deviation / safe_std
        # Fill NaN where std was 0 or near-zero (flat price → z-score = 0)
        result = result.fillna(0.0)
        return result

    @staticmethod
    def bollinger_bands(
        series: pd.Series, period: int = 20, num_std: float = 2.0
    ) -> tuple[pd.Series, pd.Series, pd.Series]:
        """Bandas de Bollinger: (upper, middle, lower)."""
        middle = series.rolling(window=period, min_periods=period).mean()
        # NaN stays NaN during warmup — fillna(0) would collapse bands to SMA,
        # causing false BB touch signals. Consumers must check for NaN.
        std = series.rolling(window=period, min_periods=period).std()
        upper = middle + num_std * std
        lower = middle - num_std * std
        return upper, middle, lower

    @staticmethod
    def rsi(series: pd.Series, period: int = 14) -> pd.Series:
        """Relative Strength Index (Wilder's smoothing)."""
        delta = series.diff()
        gain = delta.where(delta > 0, 0.0)
        loss = (-delta).where(delta < 0, 0.0)
        # Wilder's smoothing: span = 2*period - 1 (consistent with ATR)
        avg_gain = gain.ewm(span=2 * period - 1, adjust=False).mean()
        avg_loss = loss.ewm(span=2 * period - 1, adjust=False).mean()
        # Handle avg_loss=0 (pure uptrend → RSI=100) separately from initial NaN
        pure_gain = (avg_loss == 0) & (avg_gain > 0)
        rs = avg_gain / avg_loss.replace(0, np.nan)
        rsi = 100 - (100 / (1 + rs))
        rsi = rsi.where(~pure_gain, 100.0)  # Pure gains → RSI=100 (not 50)
        return rsi.fillna(50.0)  # Initial NaN (first bar) → neutral

    @staticmethod
    def momentum(series: pd.Series, period: int) -> pd.Series:
        """Momentum: retorno porcentual sobre N períodos."""
        return series.pct_change(periods=period)

    @staticmethod
    def volume_ratio(volume: pd.Series, period: int = 20) -> pd.Series:
        """Ratio de volumen actual vs media móvil."""
        avg_vol = volume.rolling(window=period, min_periods=max(period // 2, 2)).mean()
        return volume / avg_vol.replace(0, np.nan)

    @staticmethod
    def volatility_percentile(
        series: pd.Series, atr_period: int = 14, lookback: int = 100
    ) -> pd.Series:
        """Percentil de volatilidad actual dentro de ventana histórica (0-1).

        Vectorised 2026-09-04. This was `rolling(100).apply(fn, raw=False)`, which materialises a
        pandas Series per window: on the backtester it was 45 s of a 310 s run and the single largest
        indicator cost, because a 100-wide Python callable runs once per row per call. The maths below
        is the same comparison — how many of the earlier values in the window are strictly below the
        last one — done with a strided view. Identity against the old implementation is asserted in
        tests/test_indicators_vectorised.py, NaN pattern included.
        """
        returns = series.pct_change().abs()
        vol = returns.rolling(window=atr_period, min_periods=2).mean()
        v = vol.to_numpy(dtype=float)
        n = v.size
        out = np.full(n, np.nan)
        if n == 0:
            return pd.Series(out, index=series.index)

        # pandas counts a window as valid on non-NaN observations only (min_periods=10), and the
        # callable saw the FULL window including NaNs — a comparison with NaN is False either way.
        valid = (~np.isnan(v)).astype(np.int64)
        cum_valid = np.concatenate(([0], np.cumsum(valid)))
        for i in range(n):
            start = max(0, i - lookback + 1)
            if cum_valid[i + 1] - cum_valid[start] < 10:
                continue
            w = v[start:i + 1]
            denom = w.size - 1
            if denom < 1:
                out[i] = 0.5
                continue
            last = w[-1]
            if np.isnan(last):
                out[i] = 0.0
                continue
            out[i] = np.count_nonzero(w[:-1] < last) / denom
        return pd.Series(out, index=series.index)

    @staticmethod
    def _di_pair(high: pd.Series, low: pd.Series, close: pd.Series, period: int,
                 atr_val: Optional[pd.Series] = None) -> tuple[pd.Series, pd.Series]:
        """The smoothed DI+/DI- pair, unfilled. Extracted 2026-09-04.

        `adx` and `directional_indicators` computed this identical block twice, and compute_all calls
        both — three ATRs and two DM smoothings per pass where one of each will do. The arithmetic is
        untouched; tests/test_indicators_vectorised.py pins the output of every column.
        """
        plus_dm_raw = high.diff()
        minus_dm_raw = -low.diff()
        # Wilder's DM: only the larger directional move counts (ties = both zero)
        plus_dm = plus_dm_raw.where((plus_dm_raw > minus_dm_raw) & (plus_dm_raw > 0), 0.0)
        minus_dm = minus_dm_raw.where((minus_dm_raw > plus_dm_raw) & (minus_dm_raw > 0), 0.0)
        # Wilder's smoothing (equivalent to EWM with span=2*period-1)
        if atr_val is None:
            atr_val = Indicators.atr(high, low, close, period)
        smoothed_plus = plus_dm.ewm(span=2 * period - 1, adjust=False).mean()
        smoothed_minus = minus_dm.ewm(span=2 * period - 1, adjust=False).mean()
        safe_atr = atr_val.replace(0, np.nan)
        return 100 * (smoothed_plus / safe_atr), 100 * (smoothed_minus / safe_atr)

    @staticmethod
    def adx(
        high: pd.Series, low: pd.Series, close: pd.Series, period: int = 14,
        atr_val: Optional[pd.Series] = None, di: Optional[tuple] = None,
    ) -> pd.Series:
        """Average Directional Index — fuerza de tendencia (0-100).

        `atr_val` and `di` let a caller hand over work it has already done; omit them and this
        behaves exactly as it always did.
        """
        plus_di, minus_di = di if di is not None else Indicators._di_pair(high, low, close, period, atr_val)
        dx = 100 * ((plus_di - minus_di).abs() /
                     (plus_di + minus_di).replace(0, np.nan))
        return dx.ewm(span=2 * period - 1, adjust=False).mean()

    @staticmethod
    def directional_indicators(
        high: pd.Series, low: pd.Series, close: pd.Series, period: int = 14,
        atr_val: Optional[pd.Series] = None, di: Optional[tuple] = None,
    ) -> tuple[pd.Series, pd.Series]:
        """DI+ y DI- para confirmación direccional de tendencia."""
        plus_di, minus_di = di if di is not None else Indicators._di_pair(high, low, close, period, atr_val)
        return plus_di.fillna(0), minus_di.fillna(0)

    @staticmethod
    def ema_crossover(
        series: pd.Series, fast: int, slow: int
    ) -> pd.Series:
        """Señal de cruce de EMAs: +1 si fast > slow, -1 si fast < slow, 0 recién cruzó."""
        ema_fast = Indicators.ema(series, fast)
        ema_slow = Indicators.ema(series, slow)
        diff = ema_fast - ema_slow
        signal = pd.Series(0, index=series.index, dtype=float)
        signal[diff > 0] = 1.0
        signal[diff < 0] = -1.0
        return signal

    @staticmethod
    def keltner_channels(
        high: pd.Series, low: pd.Series, close: pd.Series,
        ema_period: int = 20, atr_period: int = 14, multiplier: float = 2.0
    ) -> tuple[pd.Series, pd.Series, pd.Series]:
        """Keltner Channels: (upper, middle, lower)."""
        middle = Indicators.ema(close, ema_period)
        atr_val = Indicators.atr(high, low, close, atr_period)
        upper = middle + multiplier * atr_val
        lower = middle - multiplier * atr_val
        return upper, middle, lower

    #: Every column compute_all can produce, in output order. Anything not in here is not a
    #: valid `only=` name, and asking for one is an error rather than a silently missing column.
    ALL_COLUMNS: tuple[str, ...] = (
        "sma_20", "sma_50", "ema_12", "ema_26", "atr", "std_20", "zscore",
        "bb_upper", "bb_mid", "bb_lower", "momentum_10", "momentum_20", "rsi",
        "vol_ratio", "adx", "ema_cross", "vol_pct", "plus_di", "minus_di",
        "high_20", "low_20",
    )

    @staticmethod
    def compute_all(df: pd.DataFrame, config: Optional[dict] = None,
                    only: Optional[Iterable[str]] = None) -> pd.DataFrame:
        """Calcula todos los indicadores sobre un DataFrame OHLCV.

        El DataFrame debe tener columnas: open, high, low, close, volume.
        Retorna el mismo DataFrame con columnas adicionales de indicadores.

        `only` restricts the work to the named columns. Every indicator here is independent, so a
        column computed under `only` is bit-identical to the same column computed in a full pass —
        what changes is that the other twenty are never computed at all. mean_reversion reads three
        of the twenty-one off its 1H frame and six off its 5m frame, and recomputing the rest on
        every bar was 60 % of a backtest (2026-09-04). Omit it and the behaviour is exactly as
        before. An unknown name raises rather than yielding a column that is quietly absent —
        `Series.get("x", 0)` would have swallowed a typo as a zero.
        """
        c = config or {}
        if df.empty:
            return df
        close = df["close"]
        high = df["high"]
        low = df["low"]
        volume = df["volume"]

        if only is None:
            want = set(Indicators.ALL_COLUMNS)
        else:
            want = set(only)
            unknown = want - set(Indicators.ALL_COLUMNS)
            if unknown:
                raise ValueError(f"compute_all: unknown column(s) {sorted(unknown)}")

        # Shared intermediates, computed only when something that needs them was asked for.
        need_atr = bool(want & {"atr", "adx", "plus_di", "minus_di"})
        need_di = bool(want & {"adx", "plus_di", "minus_di"})
        atr14 = Indicators.atr(high, low, close, 14) if need_atr else None
        di14 = Indicators._di_pair(high, low, close, 14, atr_val=atr14) if need_di else None
        if want & {"bb_upper", "bb_mid", "bb_lower"}:
            bb_upper, bb_mid, bb_lower = Indicators.bollinger_bands(close)
        else:
            bb_upper = bb_mid = bb_lower = None
        if need_di:
            plus_di, minus_di = Indicators.directional_indicators(high, low, close, 14, di=di14)
        else:
            plus_di = minus_di = None

        builders = {
            "sma_20": lambda: Indicators.sma(close, 20),
            "sma_50": lambda: Indicators.sma(close, 50),
            "ema_12": lambda: Indicators.ema(close, c.get("ema_fast", 12)),
            "ema_26": lambda: Indicators.ema(close, c.get("ema_slow", 26)),
            "atr": lambda: atr14,
            "std_20": lambda: Indicators.std(close, 20),
            "zscore": lambda: Indicators.zscore(close, c.get("zscore_lookback", 100)),
            "bb_upper": lambda: bb_upper,
            "bb_mid": lambda: bb_mid,
            "bb_lower": lambda: bb_lower,
            "momentum_10": lambda: Indicators.momentum(close, 10),
            "momentum_20": lambda: Indicators.momentum(close, 20),
            "rsi": lambda: Indicators.rsi(close, 14),
            "vol_ratio": lambda: Indicators.volume_ratio(volume, 20),
            "adx": lambda: Indicators.adx(high, low, close, 14, di=di14),
            "ema_cross": lambda: Indicators.ema_crossover(
                close, c.get("ema_fast", 12), c.get("ema_slow", 26)),
            "vol_pct": lambda: Indicators.volatility_percentile(close),
            "plus_di": lambda: plus_di,
            "minus_di": lambda: minus_di,
            "high_20": lambda: high.rolling(20, min_periods=20).max(),
            "low_20": lambda: low.rolling(20, min_periods=20).min(),
        }
        cols = {name: builders[name]() for name in Indicators.ALL_COLUMNS if name in want}
        if not cols:
            return df

        # ONE assignment for all of them. A loop here is still a __setitem__ per column and pandas
        # re-consolidates the block manager on every one: 71,127 of those cost 41 s of a 310 s run.
        df[list(cols)] = pd.DataFrame(
            {k: (v.to_numpy() if hasattr(v, "to_numpy") else v) for k, v in cols.items()},
            index=df.index)
        return df

File: desktop/src/lib/indicators.ts
// Indicator maths for the chart, written to the ENGINE's definitions (core/indicators.py) so the
// value on screen is the value the bot computes. No React, no store access.
//
// pandas conventions reproduced on purpose:
//  · `ewm(span, adjust=False)` is seeded with the first observation, α = 2/(span+1); Wilder's
//    smoothing is span = 2n − 1 (α = 1/n). A NaN observation keeps the previous average and lets
//    its weight decay, exactly as pandas does with ignore_na=False.
//  · `rolling(n, min_periods).std()` is the SAMPLE standard deviation (ddof = 1), and a window is
//    computed directly rather than from running sums — on a flat stretch of an 80,000 $ price a
//    running sum of squares rounds a zero variance into a small positive one.
// Every function returns an array aligned to its input, NaN where there is no value yet.
import type { Candle } from "@/stores/marketStore";

const NAN = Number.NaN;

/** pandas `Series.ewm(span=span, adjust=False).mean()` (ignore_na=False). */
export function ewm(values: number[], span: number): number[] {
  const n = values.length;
  const out = new Array<number>(n).fill(NAN);
  const alpha = 2 / (span + 1);
  const decay = 1 - alpha;
  let avg = NAN;
  let oldWt = 1;
  for (let i = 0; i < n; i++) {
    const cur = values[i];
    const obs = Number.isFinite(cur);
    if (Number.isFinite(avg)) {
      oldWt *= decay;
      if (obs) {
        if (avg !== cur) avg = (oldWt * avg + alpha * cur) / (oldWt + alpha);
        oldWt = 1;
      }
    } else if (obs) {
      avg = cur;
    }
    out[i] = avg;
  }
  return out;
}

/** pandas `rolling(n, min_periods).mean()`; NaN inside the window is skipped, as pandas does. */
export function rollingMean(values: number[], n: number, minPeriods = n): number[] {
  const out = new Array<number>(values.length).fill(NAN);
  for (let i = 0; i < values.length; i++) {
    let sum = 0;
    let count = 0;
    for (let j = Math.max(0, i - n + 1); j <= i; j++) {
      const v = values[j];
      if (Number.isFinite(v)) { sum += v; count++; }
    }
    if (count >= minPeriods && count > 0) out[i] = sum / count;
  }
  return out;
}

/** pandas `rolling(n, min_periods).std()` — sample standard deviation (ddof = 1). */
export function rollingStd(values: number[], n: number, minPeriods = n): number[] {
  const out = new Array<number>(values.length).fill(NAN);
  for (let i = 0; i < values.length; i++) {
    let sum = 0;
    let count = 0;
    const from = Math.max(0, i - n + 1);
    for (let j = from; j <= i; j++) {
      const v = values[j];
      if (Number.isFinite(v)) { sum += v; count++; }
    }
    if (count < Math.max(minPeriods, 2)) continue;
    const mean = sum / count;
    let ss = 0;
    for (let j = from; j <= i; j++) {
      const v = values[j];
      if (Number.isFinite(v)) ss += (v - mean) * (v - mean);
    }
    out[i] = Math.sqrt(ss / (count - 1));
  }
  return out;
}

function rollingExtreme(values: number[], n: number, pick: (a: number, b: number) => number): number[] {
  const out = new Array<number>(values.length).fill(NAN);
  for (let i = n - 1; i < values.length; i++) {
    let acc = NAN;
    let count = 0;
    for (let j = i - n + 1; j <= i; j++) {
      const v = values[j];
      if (!Number.isFinite(v)) continue;
      acc = count ? pick(acc, v) : v;
      count++;
    }
    if (count === n) out[i] = acc;
  }
  return out;
}

export const rollingMax = (values: number[], n: number) => rollingExtreme(values, n, Math.max);
export const rollingMin = (values: number[], n: number) => rollingExtreme(values, n, Math.min);

/** `Indicators.sma` — rolling mean with min_periods = n. */
export const sma = (values: number[], n: number) => rollingMean(values, n, n);

/** `Indicators.ema` — `ewm(span=n, adjust=False)`, defined from the first bar. */
export const ema = (values: number[], n: number) => ewm(values, n);

/** `Indicators.bollinger_bands` — SMA n ± k · sample σ, NaN through the warm-up. */
export function bollinger(closes: number[], n = 20, k = 2): { upper: number[]; mid: number[]; lower: number[] } {
  const mid = rollingMean(closes, n, n);
  const sd = rollingStd(closes, n, n);
  const upper = mid.map((m, i) => m + k * sd[i]);
  const lower = mid.map((m, i) => m - k * sd[i]);
  return { upper, mid, lower };
}

/** `high_20` / `low_20` — the Donchian channel the engine computes on this frame. */
export function donchian(candles: Candle[], n = 20): { upper: number[]; lower: number[]; mid: number[] } {
  const upper = rollingMax(candles.map((c) => c.high), n);
  const lower = rollingMin(candles.map((c) => c.low), n);
  const mid = upper.map((u, i) => (u + lower[i]) / 2);
  return { upper, lower, mid };
}

/** `Indicators.rsi` — Wilder's smoothing via ewm(span = 2n − 1); a pure run of gains reads 100,
 *  a bar without a defined ratio reads 50, exactly as the engine fills it. */
export function rsi(closes: number[], n = 14): number[] {
  const gain = new Array<number>(closes.length).fill(0);
  const loss = new Array<number>(closes.length).fill(0);
  for (let i = 1; i < closes.length; i++) {
    const d = closes[i] - closes[i - 1];
    if (d > 0) gain[i] = d;
    else if (d < 0) loss[i] = -d;
  }
  const ag = ewm(gain, 2 * n - 1);
  const al = ewm(loss, 2 * n - 1);
  return ag.map((g, i) => {
    const l = al[i];
    if (!Number.isFinite(g) || !Number.isFinite(l)) return 50;
    if (l === 0) return g > 0 ? 100 : 50;
    return 100 - 100 / (1 + g / l);
  });
}

/** `Indicators.atr` — true range with Wilder's smoothing (ewm span = 2n − 1). */
export function atr(candles: Candle[], n = 14): number[] {
  const tr = candles.map((c, i) => {
    if (i === 0) return c.high - c.low;
    const pc = candles[i - 1].close;
    return Math.max(c.high - c.low, Math.abs(c.high - pc), Math.abs(c.low - pc));
  });
  return ewm(tr, 2 * n - 1);
}

/** `Indicators.adx` + `directional_indicators` — the engine's `_di_pair`, ADX with pandas' NaN
 *  handling on the DX line (a bar where +DI + −DI is zero has no DX and does not move the ADX). */
export function adx(candles: Candle[], n = 14): { adx: number[]; plusDi: number[]; minusDi: number[] } {
  const len = candles.length;
  const plusDm = new Array<number>(len).fill(0);
  const minusDm = new Array<number>(len).fill(0);
  for (let i = 1; i < len; i++) {
    const up = candles[i].high - candles[i - 1].high;
    const down = candles[i - 1].low - candles[i].low;
    if (up > down && up > 0) plusDm[i] = up;
    if (down > up && down > 0) minusDm[i] = down;
  }
  const atrv = atr(candles, n);
  const sp = ewm(plusDm, 2 * n - 1);
  const sm = ewm(minusDm, 2 * n - 1);
  const plusRaw = sp.map((v, i) => (atrv[i] > 0 ? (100 * v) / atrv[i] : NAN));
  const minusRaw = sm.map((v, i) => (atrv[i] > 0 ? (100 * v) / atrv[i] : NAN));
  const dx = plusRaw.map((p, i) => {
    const m = minusRaw[i];
    if (!Number.isFinite(p) || !Number.isFinite(m) || p + m === 0) return NAN;
    return (100 * Math.abs(p - m)) / (p + m);
  });
  return {
    adx: ewm(dx, 2 * n - 1),
    plusDi: plusRaw.map((v) => (Number.isFinite(v) ? v : 0)),
    minusDi: minusRaw.map((v) => (Number.isFinite(v) ? v : 0)),
  };
}

/** MACD on the engine's `ema_12` − `ema_26`, signal = ewm(span 9) of that line. */
export function macd(closes: number[], fast = 12, slow = 26, signalPeriod = 9): { macd: number[]; signal: number[]; hist: number[] } {
  const ef = ewm(closes, fast);
  const es = ewm(closes, slow);
  const line = ef.map((f, i) => f - es[i]);
  const signal = ewm(line, signalPeriod);
  const hist = line.map((v, i) => v - signal[i]);
  return { macd: line, signal, hist };
}

/** `Indicators.zscore` — (close − mean) / sample σ over n bars (min_periods 2); a flat window is 0. */
export function zscore(closes: number[], n = 100): number[] {
  const mean = rollingMean(closes, n, 2);
  const sd = rollingStd(closes, n, 2);
  return closes.map((c, i) => {
    const s = sd[i];
    if (!Number.isFinite(s) || s <= 1e-12 || !Number.isFinite(mean[i])) return 0;
    return (c - mean[i]) / s;
  });
}

/** `Indicators.momentum` — `pct_change(n)`, as a fraction. */
export function momentum(closes: number[], n: number): number[] {
  return closes.map((c, i) => (i >= n && closes[i - n] !== 0 ? c / closes[i - n] - 1 : NAN));
}

/** `Indicators.volume_ratio` — volume over its n-bar mean (min_periods n/2). */
export function volumeRatio(volumes: number[], n = 20): number[] {
  const avg = rollingMean(volumes, n, Math.max(Math.floor(n / 2), 2));
  return volumes.map((v, i) => (avg[i] > 0 ? v / avg[i] : NAN));
}

/** `Indicators.volatility_percentile` — where today's |return| mean sits in its last `lookback`
 *  values (0–1), including the engine's rule that fewer than ten valid values is no answer. */
export function volatilityPercentile(closes: number[], atrPeriod = 14, lookback = 100): number[] {
  const returns = closes.map((c, i) => (i > 0 && closes[i - 1] !== 0 ? Math.abs(c / closes[i - 1] - 1) : NAN));
  const vol = rollingMean(returns, atrPeriod, 2);
  const out = new Array<number>(closes.length).fill(NAN);
  for (let i = 0; i < closes.length; i++) {
    const start = Math.max(0, i - lookback + 1);
    let valid = 0;
    for (let j = start; j <= i; j++) if (Number.isFinite(vol[j])) valid++;
    if (valid < 10) continue;
    const size = i - start + 1;
    const denom = size - 1;
    if (denom < 1) { out[i] = 0.5; continue; }
    const last = vol[i];
    if (!Number.isFinite(last)) { out[i] = 0; continue; }
    let below = 0;
    for (let j = start; j < i; j++) if (vol[j] < last) below++;
    out[i] = below / denom;
  }
  return out;
}

/** Resample 1m candles to a higher timeframe client-side (shared by chart + indicator panes). */
export function resampleCandles(candles: Candle[], tfSeconds: number): Candle[] {
  if (tfSeconds <= 60 || !candles.length) return candles;
  const buckets = new Map<number, Candle>();
  for (const c of candles) {
    const key = Math.floor(c.time / tfSeconds) * tfSeconds;
    const existing = buckets.get(key);
    if (!existing) {
      buckets.set(key, { time: key, open: c.open, high: c.high, low: c.low, close: c.close, volume: c.volume });
    } else {
      existing.high = Math.max(existing.high, c.high);
      existing.low = Math.min(existing.low, c.low);
      existing.close = c.close;
      existing.volume += c.volume;
    }
  }
  return Array.from(buckets.values()).sort((a, b) => a.time - b.time);
}
